builtin/iter.mbt

```diff
@@ -781,6 +781,28 @@ pub fn collect[T](self : Iter[T]) -> Array[T] {
   result
 }
 
+///|
+/// Collects the elements of the iterator into a string.
+pub fn join(self : Iter[String], separator~ : String = "") -> String {
+  let buf = StringBuilder::new()
```

Comment:
1. there are some duplications with `Logger::write_iter`, can you factor out some common parts
2. `if separator == ""` this rarely happens, such optimization is not worth it